src/decode.c

```diff
@@ -0,0 +1,615 @@
+#include <assert.h>
+#include <stdio.h>
+
+#include "decode.h"
+
+static inline bool decode_load(struct rv_insn_t *rv_insn, uint32_t insn)
```

Comment:
I am not sure if it makes sense to give each instruction a decode function.  Since the same type of instruction  (e.g. I-type, S-type, ...) should share the same decoding logic. I think what you should do is define every type of decoding function, for example:
```
static inline bool decode_i_type(struct rv_insn_t *rv_insn, uint32_t insn)
{
    /* I-type
     *  31       20 19   15 14    12 11   7 6      0
     * | imm[11:0] |  rs1  | funct3 |  rd  | opcode |
     */
    rv_insn->imm = dec_itype_imm(insn);
    rv_insn->rs1 = dec_rs1(insn);
    rv_insn->funct3 = dec_funct3(insn);
    rv_insn->rd = dec_rd(insn);
}
```

And then you can fix the jump table or the label for compute-goto to jump to the correct types of decoding functions.

Reply:
Thanks for your reminder!